### PR TITLE
Feat/sft liquidity

### PR DESCRIPTION
**Pull request by catalinnnn**

### CODE REVIEW

File: common/modules/reserve_queue/src/reserve_queue.rs [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        queues: &mut Vec<u64>,
+        queue_id_opt: Option<u64>,
+    ) {
+        let queue_id = queue_id_opt.unwrap_or_else(|| self.get_queue_id(amount, precision));
+        let head = self.head(token_id, queue_id).get();
+
+        if head == 0 {
+            self.init_queue(token_id, queue_id, nonce, amount);
+            queues.push(queue_id);
+            queues.sort();
```
Comment (sasurobert) [inline common/modules/reserve_queue/src/reserve_queue.rs:299]: why do you sort when only one token ?

File: common/modules/reserve_queue/src/reserve_queue.rs [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        } else {
+            self.push_to_queue(token_id, queue_id, nonce, amount);
```
Comment (sasurobert) [inline common/modules/reserve_queue/src/reserve_queue.rs:301]: are you sorting on every push ?

File: common/modules/reserve_queue/src/reserve_queue.rs [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+        &self,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+        queue_id: u64,
+    ) -> SCResult<Vec<NonceAmountPair<Self::BigUint>>> {
+        let is_fungible = !self.reserve(token_id, 0).is_empty();
+
+        if is_fungible {
```
Comment (sasurobert) [inline common/modules/reserve_queue/src/reserve_queue.rs:82]: do we actually need this if's and else's ? the queue should have actually only 1 element with fungible tokens. Which might still happen on Semi Fungible that everyone will put only 1 type of token inside - like one nonce ?

File: common/modules/reserve_queue/src/reserve_queue.rs (hunk 2 above)
Comment (sasurobert) [reply common/modules/reserve_queue/src/reserve_queue.rs:301]: so when we push - the position remains the same ?

File: common/modules/reserve_queue/src/reserve_queue.rs [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        let mut queue_id = 0;
+
+        let precision_biguint = Self::BigUint::from(precision);
+        let mut amount_clone = amount.clone();
+        while amount_clone != 0 {
+            amount_clone /= &precision_biguint;
+            queue_id += 1;
+        }
+
+        queue_id
```
Comment (sasurobert) [inline common/modules/reserve_queue/src/reserve_queue.rs:281]: Is this the binary search ? which sets the actual position for one token-nonce pair ?

File: common/modules/reserve_queue/src/reserve_queue.rs [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+        token_id: &TokenIdentifier,
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+    ) {
+        let mut elem = self.reserve(token_id, nonce).get();
+        let queue = elem.queue_id;
+        elem.amount += amount;
+        let new_queue_id = self.get_queue_id(&elem.amount, precision);
```
Comment (sasurobert) [inline common/modules/reserve_queue/src/reserve_queue.rs:316]: so these pushes would keep the list always somewhat sorted ? biggest values are always first - which should be enough, not ?

File: common/modules/reserve_queue/src/reserve_queue.rs (hunk 1 above)
Comment (catalinnnn) [reply common/modules/reserve_queue/src/reserve_queue.rs:299]: Its a vector of queue ids. when a new queue is instantiated, it needs to be added to the vector, but also the vector has to remain sorted. It should not have more than say 30 elements, so should be pretty fast and cheap

File: common/modules/reserve_queue/src/reserve_queue.rs (hunk 2 above)
Comment (catalinnnn) [reply common/modules/reserve_queue/src/reserve_queue.rs:301]: When pushing a new element to an existing queue, only its tail will be changed (and the last elem pointer to next)

File: common/modules/reserve_queue/src/reserve_queue.rs (hunk 3 above)
Comment (catalinnnn) [reply common/modules/reserve_queue/src/reserve_queue.rs:82]: I used nonce 0 in order to know whether the head/tail elements are instantiated or not. Similar to NULL. May refactor this, but some magic value will be used for this check anyway

File: common/modules/reserve_queue/src/reserve_queue.rs (hunk 4 above)
Comment (catalinnnn) [reply common/modules/reserve_queue/src/reserve_queue.rs:281]: This just maps an amount (which can be whatever) to a bucket (a queue). Element with X amount will be added to a queue with respect to a configured precision

**Comment by frick1977** [comment]
Not able to claim MEX and no to sign staking and wapping signature. Total failure for me
Maiar Chrome

**Pull request closed without merging**